modules/images/image-loading-optimization/optimization.php

```diff
@@ -149,28 +149,26 @@ function ilo_optimize_template_output_buffer( string $buffer ): string {
 	$slug = ilo_get_url_metrics_slug( ilo_get_normalized_query_vars() );
 	$post = ilo_get_url_metrics_post( $slug );
 
-	$url_metrics = $post ? ilo_parse_stored_url_metrics( $post ) : array();
-
-	$needed_minimum_viewport_widths = ilo_get_needed_minimum_viewport_widths(
-		$url_metrics,
-		microtime( true ),
+	$grouped_url_metrics = new ILO_Grouped_URL_Metrics(
+		$post ? ilo_parse_stored_url_metrics( $post ) : array(),
 		ilo_get_breakpoint_max_widths(),
 		ilo_get_url_metrics_breakpoint_sample_size(),
 		ilo_get_url_metric_freshness_ttl()
 	);
 
+	$group_statuses = $grouped_url_metrics->get_group_statuses();
+
 	// Whether we need to add the data-ilo-xpath attribute to elements and whether the detection script should be injected.
-	$needs_detection = in_array(
-		true,
-		// Each array item is array{int, bool}, with the second item being whether the viewport width is needed.
-		array_column( $needed_minimum_viewport_widths, 1 ),
-		true
-	);
+	$needs_detection = false;
+	foreach ( $group_statuses as $group_status ) {
+		if ( $group_status->is_lacking() ) {
+			$needs_detection = true;
+			break;
+		}
+	}
```

Comment:
See above: Instead of this logic, why not use a class method like `is_every_group_full()` to simplify this?
```suggestion
	$needs_detection = ! $grouped_url_metrics()->is_every_group_full();
```

Reply:
See 28ee2b8